### DISCUSSION TITLE
Q: Hosting the app as a Flask compatible web app in Dataiku

### DISCUSSION THREAD

**Discussion by thondeboer** [writer/writer-framework#152] [General]
I'm a big fan of low code environments such as Tableau and Dataiku DSS and the latter has the capability to host webapps, but they only support a few specific (and one general webapp) category. Dash, Boke (and R shiny, but I want to focus on Pyhon).

There is also the "generic webapp" category which allows the user to create an HTML file, CSS, JS and a python backend, but that backend uses FLASK for its routing.

I am not very familiar with all the nitty gritty differences between WSGI and ASGI types of applications servers so I aooplogize if this is a silly question, but is there a way to host a streamsync/uvicorn app in this environment such as dataiku, where I don't have access to the NGINX configurations...

See more on dataiku webapps here for some background: https://doc.dataiku.com/dss/latest/webapps/index.html

**Comment by thondeboer** [comment]
I tried the following (thanks chatGPT)

HTML file
```html
<iframe id="myIframe"></iframe>
```

Javascript
```js
$.get(getWebAppBackendUrl('/hello'), function(data) {
    const iframe = document.getElementById('myIframe');
    iframe.contentWindow.document.open();
    iframe.contentWindow.document.write(data);
    iframe.contentWindow.document.close();
});
```

Backend
```python
import dataiku
import pandas as pd
from flask import redirect, request

import streamsync.serve
import uvicorn
import os
import threading

# Your Uvicorn ASGI app
def run_uvicorn():
    app_path = os.environ['MY_APP_PATH']
    print(app_path)
    mode = "run" # run or edit

    asgi_app = streamsync.serve.get_asgi_app(app_path, mode)

    uvicorn.run(asgi_app,
               host="127.0.0.1",
               port=5000,
               log_level="warning")

@app.route('/hello')
def first_call():
    return redirect("http://127.0.0.1:5000")


t = threading.Thread(target=run_uvicorn)
t.start()
```
(The MY_APP_PATH has the location of the main.py and ui.json etc.)

It does seem to run the hello app, but the redirection does not seem to work. This probably is too naieve a solution, since ASGI etc. using web sockets and this is not using that and probably requires more sophisticated soltuion that is beyond me ATM.

These are the logs I get form dataiky
```
[local] 2023-09-27 15:52:14,683 INFO Starting Webapp backend
[local] 2023-09-27 15:52:14,683 INFO Serving dependencies for html
[local] 2023-09-27 15:52:14,684 INFO Serving html
[local] 2023-09-27 15:52:14,684 INFO Starting backend for web app: TEST.z912iCE
[local] /DSS/data/code-envs/python/streamsync/lib/python3.10/site-packages/fastapi/openapi/models.py:55: DeprecationWarning: `general_plain_validator_function` is deprecated, use `with_info_plain_validator_function` instead.
[local]   return general_plain_validator_function(cls._validate)
[local] /DSS/data/code-envs/python/streamsync/lib/python3.10/site-packages/pydantic_core/core_schema.py:3898: DeprecationWarning: `general_plain_validator_function` is deprecated, use `with_info_plain_validator_function` instead.
[local]   warnings.warn(
[local] /DSS/data/code-envs/resources/python/streamsync/ss-hello/ss-hello-world
[local] 2023-09-27 15:52:14,977 INFO Started backend on port 45131
[local] 2023-09-27 15:52:14,986 INFO 127.0.0.1 - - [27/Sep/2023 15:52:14] "GET /__ping HTTP/1.1" 200 -
[local] 2023-09-27 15:52:15,000 INFO [34;20mFrom app log: 
[local] Stdout message during initialisation
[local] Hello world!
[local] [0m
[local] [34;20mFrom app log: 
[local] Stdout message during initialisation
[local] Hello world!
[local] [0m
[local] 2023-09-27 15:52:15,000 INFO [34;20mFrom app log: 
[local] Stdout message during initialisation
[local] Hello world!
[local] [0m
[local] 2023-09-27 15:52:15,244 INFO 127.0.0.1 - - [27/Sep/2023 15:52:15] "GET /__ping HTTP/1.1" 200 -
[local] 2023-09-27 15:52:15,499 INFO 127.0.0.1 - - [27/Sep/2023 15:52:15] "[32mGET /hello HTTP/1.1[0m" 302 -
```

I do see to get the 302 for the redirect, but nothing shows up in the iframe...

**Comment by thondeboer** [comment]
Also tried this as JS to na avail...
```js
document.getElementById('myIframe').src = "http://127.0.0.1:5000";
```

**Comment by thondeboer** [comment]
Well...I made it work!
In the end it was incredible simply to make it work, not sure why I struggled so much before :)

There was a small added complication, since I am running dataiku on a docker container, but this is all I needed to do

1) Ensure the port (5000) in my case, is open on docker (So -p 5000:5000 for docker command)
2) Create a simple one line HTML file with the hard code iP address for my local machine
```html
<iframe src="http://192.168.4.183:5000"></iframe>
```
3) Run the uvicorn app in a separate thread
```python
import dataiku
import pandas as pd
from flask import redirect, request

import streamsync.serve
import uvicorn
import os
import threading

# Your Uvicorn ASGI app
def run_uvicorn():
    app_path = os.environ['MY_APP_PATH']
    print(app_path)
    mode = "run" # run or edit

    asgi_app = streamsync.serve.get_asgi_app(app_path, mode)

    uvicorn.run(asgi_app,
               host="0.0.0.0",
               port=5000,
               log_level="warning")

t = threading.Thread(target=run_uvicorn)
t.start()
```

And that's it! Now I can run my app inside dataiku as a webapp, allowing all the benefits of dataiky and streamsync!

**Comment by ramedina86** [comment]
Thanks for conducting and more importantly sharing this research.

@FabienArcellier also worked on a way to speed up deployment of Streamsync in a cloud environment via Gitpod (quite impressive) https://github.com/FabienArcellier/blueprint-streamsync

Looks like these options would work well for the official documentation.
